### PR TITLE
[QA-625] - Enable users and user groups to be able to be tagged in messages

### PR DESCRIPTION
**Pull request by KellenKincaid**
**NOTE**: _The vast majority of the changed files are auto-generated node_modules files, resulting from me installing Typescript. Ignore those._

[ WHAT IS THE PURPOSE OF THIS PR? ]
Enable the Slack Notifier orb to be able tag users and user groups in messages.

[ HOW DOES THIS WORK? ]
First, all Vouch user IDs and user group IDs are collected via a Slack API call.
The collection of these IDs requires a Bot User OAuth Token from a Slack App (Notifier).
Once collected, these IDs are stored in a temporary file (_slackIdLookupTable.txt_).

When a user is tagged in a message (like this: _@user_name_) the tagged user is replaced with _\<@the_user_id\>_ which is the syntax required to have Slack tag them in the message.
Similarly, when a user group is tagged in a message (like this: _!user_group_) the tagged user group is replaced with _\< !subteam^the_user_group_id \>_ which is the syntax required to have Slack tag the user group in a message.

[ WHICH TEAM IS THIS WORK FOR? ]
Quality (SDET)

[ JIRA CARD LINKS ]
https://vouchinc.atlassian.net/browse/QA-625
https://vouchinc.atlassian.net/browse/QA-614

### CODE REVIEW

**Comment by snowinferno** [comment]
I'm not overly familiar with ORB best practices, but from a node/typescript standpoint it's typically bad practice to commit `node_modules`. `yarn.lock` will cause the same version of the packages to be installed when running `yarn install` with some exceptions I won't detail here. In CI environments, `yarn install --frozen-lockfile` will strictly honor the lock file without updating any versions.

If node modules need to be installed for parts of the CI process, there should be a predicate step or command that runs `yarn install --frozen-lockfile` to install the node dependencies before any step or command that tries to use them.

**Review by snowinferno — CHANGES_REQUESTED** [review]
TL;DR of my comments:

1. use `yarn install --frozen-lockfile`. There have been times when a module developer releases a breaking change in a _**patch version**_ which `yarn` and `yarn install` will install given the entries in the `package.json`.
2. Unless there is a compelling reason to commit `node_modules`, add `node_modules` to `.gitignore` and delete `node_modules` from the repo. It makes the repo unnecessarily large and adds noise to PRs any time there is a change in dependencies.

Nitpicks, not requiring the below changes for approval:

By default, the environment gets loaded into `process.env` and can be read from there. No matter what the env variable is it will be provided on `process.env` as a string. Assigning values into `process.env` will convert any value to a string. This may reduce the number of arguments being supplied to the TS functions if it's just passing in values that are also on the environment. By default `execSync` provides the same values from `process.env` to the new child process.

File I/O can be done all within node without having to spawn a new process and spin up a new bash shell to execute the bash script. Node built-in I/O is done on a protected separate thread which uses OS interfaces to perform the I/O. Having said that, I don't think there is significant difference in performance whether node built-in I/O is used or a new process is spawned to do the I/O through bash.

File: src/examples/example.yml [hunk 1]
```diff
@@ -51,11 +51,26 @@ usage:
             header: TESTING SLACK-NOTIFICATION ORB - Failed step!
             channel-webhook-environment-variables: TEST_SLACK_WEBHOOK_1 TEST_SLACK_WEBHOOK_2
             additional-text: "This step is supposed to _fail_ for the sake of demonstration."
+    test-tagging-users-and-groups:
+      docker:
+        - image: cimg/base:current
```
[the remark this answers is not in the document]
Comment (KellenKincaid) [reply src/examples/example.yml]: Good catch.
Changed `cimg/base:current` to `cimg/node:current` where appropriate.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,9 @@
...
 orb.yml
-.DS_Store
\ No newline at end of file
+.DS_Store
+slackIds.json
+slackUserInfo.json
+slackGroupInfo.json
+deleteme*
+temporary*
+**/dist
```
[the remark this answers is not in the document]
Comment (KellenKincaid) [reply .gitignore]: Big brain fart.
Removed `node_modules` from the remote repo and added it to the `.gitignore`.

File: src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import { readFile } from 'fs'
+import { exec } from 'child_process'
+import { SlackUser, SlackGroup, SlackUsersResponseObject, SlackGroupsResponseObject } from './interfaces'
+
+export async function PopulateLookupTable(_lookupTable:string[], _writeLookupTableShellScriptFilePath:string, _lookupTableFilePath:string, _getSlackUserShellScriptFilePath:string, _slackUserInfoFilePath:string, _slackGroupInfoFilePath:string) {
+    await AddUserDataToLookupTable(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath, _slackUserInfoFilePath);
+    await AddUserGroupDataToLookupTable(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath, _slackGroupInfoFilePath);
+}
+
+async function AddUserDataToLookupTable(_lookupTable:string[], _writeLookupTableShellScriptFilePath:string, _lookupTableFilePath:string, _slackUserInfoFilePath:string) {
```
Comment (snowinferno) [inline src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts:10]: How important is it to wait for `readFile` to finish reading the file in and `exec` to finish writing the file back out before anything else happens in the code that calls this?

As it stands, the code calling this method will continue without waiting for the file to be fully read and without waiting for the `exec` call to finish.

There are a few options depending on which asynchronous activities need to be completed before continuing code execution.

File: src/scripts/slack_id_lookup_table_scripts/generate_slack_id_lookup_table.sh [hunk 4]
```diff
@@ -1,9 +1,5 @@
 #!/bin/bash
 
-getSlackUserShellScriptFilePath='src/scripts/slack_id_lookup_table_scripts/gather_slack_data.sh'
-slackUserInfoFilePath='slackUserInfo.json'
-slackGroupInfoFilePath='slackGroupInfo.json'
-slackIdLookupGeneratorFilePath='src/scripts/slack_id_lookup_table_scripts/generate_slack_id_lookup_table.ts'
-
-bash "${getSlackUserShellScriptFilePath}" "${!SLACK_BOT_TOKEN}" "${slackUserInfoFilePath}" "${slackGroupInfoFilePath}"
-npx ts-node "${slackIdLookupGeneratorFilePath}"
+mkdir "${SLACK_DATA_DIRECTORY_PATH}"
```
Comment (snowinferno) [inline src/scripts/slack_id_lookup_table_scripts/generate_slack_id_lookup_table.sh:3]: If a user is allowed to specify a deep path, say something like `tmp/some/folder/slackData`, `mkdir -p` would make sure the full path is created if any segments of the path are missing.

File: src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts [hunk 5]
```diff
@@ -1,107 +1,73 @@
...
+import { execSync } from 'child_process'
 import { SlackUser, SlackGroup, SlackUsersResponseObject, SlackGroupsResponseObject } from './interfaces'
 
-export async function PopulateLookupTable(_lookupTable:string[], _writeLookupTableShellScriptFilePath:string, _lookupTableFilePath:string, _getSlackUserShellScriptFilePath:string, _slackUserInfoFilePath:string, _slackGroupInfoFilePath:string) {
-    await AddUserDataToLookupTable(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath, _slackUserInfoFilePath);
-    await AddUserGroupDataToLookupTable(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath, _slackGroupInfoFilePath);
+export function PopulateLookupTable(_lookupTable:string[], _writeLookupTableShellScriptFilePath:string, _lookupTableFilePath:string, _getSlackUserShellScriptFilePath:string, _slackUserInfoFilePath:string, _slackGroupInfoFilePath:string):void {
+    AddUserDataToLookupTable(_lookupTable, _slackUserInfoFilePath);
+    AddUserGroupDataToLookupTable(_lookupTable, _slackGroupInfoFilePath);
+    WriteLookupTableToFile(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath);
```
Comment (snowinferno) [inline src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts:8]: A try/catch block is recommended here in case there are any errors reading the file, parsing the JSON, or executing the bash script to write the file. Without catching the exception, the process will exit out with an `UncaughtErrorException` detailing the uncaught error.

Alternatively a try/catch block can be put around each call to `readFileSync`, `JSON.parse`, and `execSync` calls to get finer control of error output.

```suggestion
    try {
        AddUserDataToLookupTable(_lookupTable, _slackUserInfoFilePath);
        AddUserGroupDataToLookupTable(_lookupTable, _slackGroupInfoFilePath);
        WriteLookupTableToFile(_lookupTable, _writeLookupTableShellScriptFilePath, _lookupTableFilePath);
    } catch (e) {
        console.error('Something went wrong!', e);
    }
```

**Review by snowinferno — APPROVED** [review]
Looks good to me!

I left a couple of recommendations but I don't think they need to hold up approval.

**Review by snowinferno — APPROVED** [review]

File: src/scripts/slack_id_lookup_table_scripts/generate_slack_id_lookup_table.sh (hunk 4 above)
Comment (KellenKincaid) [reply src/scripts/slack_id_lookup_table_scripts/generate_slack_id_lookup_table.sh:3]: Ahh good to know! I'll add this change in a future PR

File: src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts (hunk 5 above)
Comment (KellenKincaid) [reply src/scripts/slack_id_lookup_table_scripts/lookup_table_functions.ts:8]: Great idea, I'll add a try-catch block to each critical point with a detailed error message.
